Added docs for TS SDK. (#2334)
wasp-lang/wasp: web/docs/general/wasp-ts-config.md

This feature is currently in early preview and we are actively working on it.
:::

In Wasp, you normally define/configure the high level of your app (pages, routes, queries, actions, auth, ...) in a `main.wasp` file in the root of your project. In `main.wasp` you write in Wasp's DSL (domain-specific language), which is a simple configuration language similar to JSON but smarter.

In Wasp 0.15, we introduce **Wasp TS config**, an alternative way to define the high level of your app via `main.wasp.ts`! Although it looks extremely similar to how you would do it in `main.wasp`, the difference is that you write in TypeScript, not in Wasp's DSL.

This is an **early preview** feature, meaning it is quite rough and unpolished and not yet where it could be, but it does work, we think it is pretty cool already, and you can try it out now. If you do, please share your feedback and any ideas with us on our [GitHub](https://github.com/wasp-lang/wasp) or <DiscordLink />, this is crucial for us to be able to shape this feature in the best possible way!

## Motivation

- Out-of-the-box support in all editors.
- Less maintenance on our side.
- More flexibility for you while writing the config.

> - It makes it easy to support splitting the Wasp config into multiple files.

Author: Yes but was talking about the future hm, not the present! And that is lost now.

- Lays down a good foundation for the Full Stack Modules (FSM) that are a part of our future plans.

## How to switch from Wasp DSL config to Wasp TS config

1. Go into your Wasp project that you want to switch to Wasp TS config (or create a new Wasp project if you want to try it out like that). Make sure you are on Wasp >= 0.15 and your project is working.

> 2. Rename `tsconfig.json` file to `tsconfig.src.json` and add `"include": ["src"]` entry to it:

Author: ```suggestion
2. Rename `tsconfig.json` file to `tsconfig.src.json` and add `"include": ["src"]` entry to the top level (after `compileOptions`, not in them!):
```

    }
    ```

3. Create new `tsconfig.json` file with the following content:

    ```json title="tsconfig.json"
    {
      "files": [],
      "references": [
        { "path": "./tsconfig.src.json" },
        { "path": "./tsconfig.wasp.json" }
      ]
    }
    ```

4. Create new `tsconfig.wasp.json` file with the following content:

    ```json title="tsconfig.wasp.json"
    {
      "compilerOptions": {
        "skipLibCheck": true,
        "target": "ES2022",
        "isolatedModules": true,
        "moduleDetection": "force",

        // linting
        "strict": true,
        "noUnusedLocals": true,
        "noUnusedParameters": true,

        "module": "NodeNext",
        "noEmit": true,

        "lib": ["ES2023"],
      },
      "include": ["main.wasp.ts"]
    }
    ```

5. In your `package.json`, add `"type": "module"` entry at the top level:

    ```json title="package.json"
    {
      "type": "module",
      ...
    }
    ```

6. Rename `main.wasp` file to `main.wasp.old`: we want to keep it for now so we can look at it while writing `main.wasp.ts` based on it.
7. Run `wasp clean` and `rm package-lock.json` to ensure we have a clean slate to start from.

> 8. Run `wasp ts-setup` to do the basic setup for Wasp TS config. It will also add `wasp-config` package to `devDependencies` in your `package.json`.

Reviewer: ```suggestion
8. Run `wasp ts-setup`. This command will add the `wasp-config` package to your `package.json`'s `devDependencies`.
```

> 9. Create an empty `main.wasp.ts` file and rewrite your `main.wasp.old` in it but in TypeScript.

Reviewer: ```suggestion
9. Create a file called `main.wasp.ts` and translate the config from `main.wasp.old` (written in Wasp) into `main.wasp.ts` (written in TypeScript).
```

>    Check out the [reference main.wasp.ts file](#reference-main-wasp-ts-file) below for details on what the TypeScript API for configuring Wasp looks like.

Reviewer: ```suggestion
   Check out the [reference main.wasp.ts file](#reference-main-wasp-ts-file) below for more details on the new TypeScript API.
```

> In short, you will want to import the `App` from `wasp-config`, use it to define parts of your web app like `auth`, `pages`, `query`, `api`, ..., and at the end do the `default export app`.

Reviewer: ```suggestion
In short, you'll have to:
  1. Import `App` from `wasp-config`
  2. Create a new `app` object with `new App()`.
  3. Use the `app` object to define parts of your web app like `auth`, `pages`, `query`, `api`...
  4. Export the `app` from your file using a default export.
```

>    You can manually do the rewrite using the reference file and TS types as guides (IDE support should work for you in `main.wasp.ts`), or you can (and we recommend it!) give the reference main.wasp.ts file to the LLM of your choice and tell it to rewrite your `main.wasp` while following the format in the reference file: we had great results with this!

Reviewer: ```suggestion
   You can translate your config manually with the help of the reference file (see below) and your IDE (IDE support should work in `main.wasp.ts`). Another option (which we recommend!) is giving the reference `main.wasp.ts` file to the LLM of your choice and asking it to translate your `main.wasp` file by following the reference file: We had great results doing this!
```

> 10. Run `wasp start` to run your app! If you got it all right, it should work the same as it worked before: the only difference is that it is now reading the Wasp config from `main.wasp.ts` instead of `main.wasp`.

Reviewer: ```suggestion
10. Run `wasp start` to run your app! If you got everything right, your app should work exactly like it did before. The only difference is that it's now reading the Wasp config from `main.wasp.ts` instead of `main.wasp`.
```

    :::tip
      Don't forget to have the database running or do the db migrations if needed, as you would normally when running your app in development.
    :::

> 11. That is it, you are now using Wasp TS config! You can delete `main.wasp.old` file now if you still have it around.

Reviewer: ```suggestion
11. That's it! You're now using the Wasp TS config! If everything works, you can delete the `main.wasp.old` file.
```

Got stuck on any of these steps? Let us know in our <DiscordLink /> and we will help!

> ## What next?

Reviewer: ```suggestion
## What to do next?
```

### Experiment

> Play with the Wasp TS config, see if you can improve it in any way, and get a feel for it. Some ideas to experiment with:

Reviewer: ```suggestion
Play with the Wasp TS config, get the feel of it, and see if you can find ways to improve it. Here are some ideas you can experiment with:
```

> - How would you go about reducing the boilerplate in `main.wasp.ts` file? Helper functions, loops?

Reviewer: ```suggestion
- How would you reduce the boilerplate in `main.wasp.ts` file? Helper functions, loops?
```

> - Can you imagine a better API / abstractions? If you do, what would it look like? Maybe you can even implement it on top of our API?

Reviewer: ```suggestion
- Can you imagine a better API or better abstractions? If you can, what would that look like? Perhaps you can even implement it on top of our API?
```

> - Give a try at implementing your own file-based routing if that is what you like: you are now in Turing complete language and have access to the disk!

Reviewer: ```suggestion
- Try implementing file-based routing if that's what you like. You are now working in a Turing-complete language with full access to your system!
```

- Surprise us!

### Feedback

> Whatever you end up doing, we would love it if you would let us know how it was and show us what you did.

Reviewer: ```suggestion
Whatever you end up doing, we would love it if you could let us know what you thought and show us what you did.
```

> We do have some immediate ideas of our own about what we want to improve, but we want to hear what you thought of, what you liked or disliked, or what you came up with. Even if you just found it all good, or just a single thing you didn't or did like, that is also valuable feedback and we would love to hear it!

Reviewer: ```suggestion
We have some ideas about what we want to improve, but we want to hear what you thought of the new TS SDK: what you liked, disliked, what you came up with, etc.
Even if you liked everything, or if there's just a single small thing you didn't like, we want to hear it. It's all very valuable feedback.
```

> Let us know either on our [GitHub](https://github.com/wasp-lang/wasp) or even better in our <DiscordLink />.

Reviewer: ```suggestion
Let us know on our [GitHub](https://github.com/wasp-lang/wasp) or, even better, in our <DiscordLink />.
```
